lib/active_merchant/billing/gateways/rapyd.rb

```diff
@@ -0,0 +1,239 @@
+          end
+        else
+          commit(:post, 'payments', post)
+        end
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+        add_invoice(post, money, options)
+        add_payment(post, payment, options)
+        add_address(post, payment, options)
+        post[:capture] = false
+        commit(:post, 'payments', post)
+      end
+
+      def capture(money, authorization, options = {})
+        post = {}
+        commit(:post, "payments/#{authorization}/capture", post)
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        post[:payment] = authorization
+        add_invoice(post, money, options)
+        commit(:post, 'refunds', post)
+      end
+
+      def void(authorization, options = {})
+        post = {}
+        commit(:delete, "payments/#{authorization}", post)
+      end
+
+      # Gateway returns an error if trying to run a $0 auth as invalid payment amount
+      # Gateway does not support void on a card transaction and refunds can only be done on completed transactions (such as a purchase). Authorize transactions are considered 'active' and not 'complete'  until they are captured.
+      def verify(credit_card, options = {})
+        MultiResponse.run do |r|
+          r.process { purchase(100, credit_card, options) }
+          r.process { refund(100, r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Access_key: )\w+), '\1[FILTERED]').
+          gsub(%r(("number\\?":\\?")\d+), '\1[FILTERED]').
+          gsub(%r(("cvv\\?":\\?")\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      def payment_is_ach?(options)
+        return true if options[:type].include?('bank')
+      end
+
+      def add_address(post, creditcard, options)
+        return unless address = options[:address]
+
+        post[:address] = {}
+        # name and line_1 are required at the gateway
+        post[:address][:name] = address[:name] if address[:name]
+        post[:address][:line_1] = address[:address1] if address[:address1]
+        post[:address][:line_2] = address[:address2] if address[:address2]
+        post[:address][:city] = address[:city] if address[:city]
+        post[:address][:state] = address[:state] if address[:state]
+        post[:address][:country] = address[:country] if address[:country]
+        post[:address][:zip] = address[:zip] if address[:zip]
+        post[:address][:phone_number] = address[:phone] if address[:phone]
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money).to_f.to_s
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def add_payment(post, payment, options)
+        type = options[:type]
+
+        post[:payment_method] = {}
+        post[:payment_method][:fields] = {}
+        post[:payment_method][:type] = type
```

Comment:
Stylistic nitpick: I know these lines are common to both `add_credit_card` and `add_ach` but I'm wondering if they more appropriately belong in each of those methods, so that `add_payment` simply switches on `payment_is_card?` or `payment_is_ach?`.

If you don't like that idea, it might be nice instead to include a guard clause in `add_credit_card` and `add_ach`, like:
```
return unless (pm_fields = post[:payment_method][:fields])

pm_fields[:number] = payment.number
etc...
```

Reply:
I think I revised this to be what you suggested, but let me know if I didn't get it quite right 😅

Reply:
Nailed it 💪